src/typecheck.rs

```diff
@@ -1,11 +1,1384 @@
+                    expected: Type::BoolT,
+                    actual: cond_type,
+                });
+            };
+
+            // Type check both branches
+            typeck_block(true_branch, ctx)?;
+            typeck_block(false_branch, ctx)?;
+            Ok(())
+        }
+    }
+}
+
+/// Typecheck a block of statements in the given environment using typeck_stmt
+pub fn typeck_block(block: &Block, ctx: &TypeCtx) -> Result<(), TypeError> {
+    // type check each stmt in the vector sequence is ok
+    for stmt in block {
+        typeck_stmt(stmt, ctx)?;
+    }
+    Ok(())
+}
+
+///Typecheck a function using the given environment and function environment
+pub fn typeck_function(fun: &Function, ctx: &TypeCtx) -> Result<(), TypeError> {
+    let mut fun_env = ctx.env.clone();
+
+    for (param, param_type) in &fun.params {
+        fun_env.insert(*param, param_type.clone());
+    }
+
+    let fun_ctx = TypeCtx {
+        env: fun_env,
+        funcs: ctx.funcs,
+    };
+
+    let e = typeck_expr(&fun.body, &fun_ctx)?;
+
+    //check that body is the same type as the return type, otherwise return error
+    if e == fun.ret_typ {
+        Ok(())
+    } else {
+        let t = fun.ret_typ.clone();
+        Err(TypeError::TypeMismatch {
+            expected: t,
+            actual: e,
+        })
+    }
+}
+
+/// Type check a program
+pub fn typecheck_program(program: &Program) -> Result<(), TypeError> {
+    // Create initial context
+    let ctx = TypeCtx {
+        env: program.locals.clone(),
+        funcs: &program.helpers,
+    };
+
+    // Check that all helper functions are well-typed
+    for (func_name, function) in &program.helpers {
+        if let Err(err) = typeck_function(function, &ctx) {
+            println!("Type error in function '{}': {}", func_name, err);
+            return Err(err);
+        }
+    }
+
+    // Type check all statements in the block
+    if let Err(err) = typeck_block(&program.start, &ctx) {
+        println!("Type error in block: {}", err);
+        return Err(err);
+    }
 
-enum TypeError {
-    TypeMismatch,
+    // Type check the action handler
+    let (input_var, action_block) = &program.action;
+
+    // Add the input variable to the environment as a symbol type if it exists
+    let mut action_env = ctx.env.clone();
+    if let Some(var) = input_var {
+        action_env.insert(*var, Type::SymT);
+    }
+
+    // Type check the action block
+    typeck_block(
+        action_block,
+        &TypeCtx {
+            env: action_env,
+            funcs: ctx.funcs,
+        },
+    )?;
+
+    // Type check the accept condition (must be a boolean)
+    let final_type = typeck_expr(&program.accept, &ctx)?;
+    if final_type != Type::BoolT {
+        return Err(TypeError::TypeMismatch {
+            expected: Type::BoolT,
+            actual: final_type,
+        });
+    }
+
+    Ok(())
 }
 
-fn typeck(program: &Program) -> Result<(), TypeError> {
-    todo!()
+#[cfg(test)]
+mod tests {
```

Comment:
I'll review the tests after you fix the issues above. I left a few quick notes below.